Allow pattern matching `null` against pointer types when the pointer types contain nested type parameters

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/49639.

I've gone through other references to `IsNonNullableValueType`, and none of them need updating. The rest of the uses center around type parameter constraint checking, which function pointers cannot be, and lifted conversion operators, where the underlying value type is being checked with this method, and function pointers cannot be that underlying type.

333fred commented:
@dotnet/roslyn-compiler for review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -345,7 +345,8 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
                     {
-                        if (inputType.IsNonNullableValueType())
+                        // Function pointers are value types, but they can be assigned null, so they can be matched against null.
+                        if (inputType.IsNonNullableValueType() && inputType.TypeKind != TypeKind.FunctionPointer)
```
> if (inputType.IsNonNullableValueType() && inputType.TypeKind != TypeKind.FunctionPointer) [](start = 23, length = 90)

Are we not going through this code path for non-generic case?

333fred replied:
Nope. The first `if` in the function checks for type parameters, and goes down a different path for them.

---
In reply to: [541360178](https://github.com/dotnet/roslyn/pull/49915#discussion_r541360178) [](ancestors = 541360178)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -345,7 +345,8 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
                     {
-                        if (inputType.IsNonNullableValueType())
+                        // Function pointers are value types, but they can be assigned null, so they can be matched against null.
+                        if (inputType.IsNonNullableValueType() && inputType.TypeKind != TypeKind.FunctionPointer)
```
>inputType.TypeKind != TypeKind.FunctionPointer [](start = 66, length = 46)

It looks like we have the same issue with regular pointers, I think we should use IsPointerOrFunctionPointer helper here instead.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -11176,6 +11176,45 @@ public void InvalidReturnInCompoundAssignment()
+{
+    test<int>(null);
+
+    static void test<T>(delegate*<T, void> f)
```
>static void test<T>(delegate*<T, void> f) [](start = 4, length = 41)

Perhaps should cover non-generic case as well for the purpose of catching possible regression in the future.

333fred replied:
See `FunctionPointerTypePatternIsNull`: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs#L2340](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs#L2340).

---
In reply to: [541365340](https://github.com/dotnet/roslyn/pull/49915#discussion_r541365340) [](ancestors = 541365340)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -11176,6 +11176,45 @@ public void InvalidReturnInCompoundAssignment()
+using System.Threading.Tasks;
+unsafe
+{
+    test<int>(null);
```
>test<int>(null); [](start = 4, length = 16)

For completeness, let's cover non-null value too.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -11176,6 +11176,45 @@ public void InvalidReturnInCompoundAssignment()
         }
 
+        [Fact, WorkItem(49639, "https://github.com/dotnet/roslyn/issues/49639")]
+        public void CompareToNullWithNestedUnconstrainedTypeParameter()
```
>CompareToNullWithNestedUnconstrainedTypeParameter [](start = 20, length = 49)

Please add similar test for regular pointers.

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [743462458](https://github.com/dotnet/roslyn/pull/49915#issuecomment-743462458) [](ancestors = 743462458)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:
```diff
@@ -9539,6 +9539,92 @@ .maxstack  2
         }
 
+        [Fact, WorkItem(49639, "https://github.com/dotnet/roslyn/issues/49639")]
+        public void CompareToNullWithNestedUnconstrainedTypeParameter()
```
>CompareToNullWithNestedUnconstrainedTypeParameter [](start = 20, length = 49)

I assume a non-generic test is covered elsewhere.

333fred replied:
Yes: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs#L4463](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs#L4463)

---
In reply to: [541420114](https://github.com/dotnet/roslyn/pull/49915#discussion_r541420114) [](ancestors = 541420114)

AlekseyTs reviewed (approved):
LGTM (commit 2). Please ensure to adjust the title to reflect the actual scope of the change when merging.

333fred commented:
@dotnet/roslyn-compiler for a second review of this small change.

jaredpar commented on src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:
```diff
@@ -9539,6 +9539,92 @@ .maxstack  2
+  IL_0014:  ret
+}
+");
+        }
```
Can we add a test where we pattern match against a field which is a pointer and verify that this errors if it's done outside of `unsafe` code? The use of fields as pointers and it's interactions with the `unsafe` context has been a trouble area in the past.

333fred commented:
> Please ensure to adjust the title to reflect the actual scope of the change when merging.

Done.
